sdks/go/pkg/beam/core/runtime/exec/fn_test.go

```diff
@@ -195,6 +196,50 @@ func TestInvoke(t *testing.T) {
 	}
 }
 
+func TestRegisterCallback(t *testing.T) {
+	bf := bundleFinalizer{
+		callbacks:         []bundleFinalizationCallback{},
+		lastValidCallback: time.Now(),
+	}
+	testVar := 0
+	bf.RegisterCallback(500*time.Minute, func() error {
+		testVar += 5
+		return nil
+	})
+	bf.RegisterCallback(2*time.Minute, func() error {
+		testVar = 25
+		return nil
+	})
+	callbackErr := errors.New("Callback error")
+	bf.RegisterCallback(2*time.Minute, func() error {
+		return callbackErr
+	})
+
+	// We can't do exact equality since this relies on real time, we'll give it a broad range
+	if bf.lastValidCallback.Before(time.Now().Add(400*time.Minute)) || bf.lastValidCallback.After(time.Now().Add(600*time.Minute)) {
+		t.Errorf("RegisterCallback() lastValidCallback set to %v, want about 500 minutes", bf.lastValidCallback)
+	}
+	if got, want := len(bf.callbacks), 3; got != want {
+		t.Fatalf("RegisterCallback() called twice, got %v callbacks, want %v", got, want)
```

Comment:
Nit: Looks like a lot of these error messages were written with the wrong number of callbacks in mind (like here says it was called twice when it was called three times). For the most part though, I think you can avoid hardcoding that into the error message entirely. Something like "callbacks in bundleFinalizer does not match number of calls to RegisterCallback(), got %v callbacks, want %v". Wordier but doesn't need to be changed whenever the test changes.

Reply:
I like it, good call